experimental/DoubleAndHyperComplexes/src/Objects/tensor_products.jl

```diff
@@ -196,3 +196,186 @@ function tensor_product(list::Vector{T}) where {T<:ComplexOfMorphisms}
+  #      original method for `tensor_product` so that the anticipated output is produced
+  #
+  #   2) If that does not work, you may also overwrite the whole method for production 
+  #      of the tensor products here.
+  result, mult_map = tmp_res
+  fac.mult_map_cache[I] = mult_map
+  return result
+end
+
+# By default, we assume that `tensor_product` produces the correct output
+function _tensor_product(v::Any...)
+  return tensor_product(v...)
+end
+
+# If not, we can wrap it here so that the correct output is produced
+function _tensor_product(v::ModuleFP...)
+  return tensor_product(v...; task=:with_map)
+end
+
+function can_compute(fac::HCTensorProductChainFactory, c::AbsHyperComplex, I::Tuple)
+  i = collect(I)
+
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  return all(k->can_compute_index(fac.factors[k], Tuple(j[k])), 1:length(j))
+end
+
+### Production of the morphisms 
+struct HCTensorProductMapFactory{MorphismType} <: HyperComplexMapFactory{MorphismType}
+  factors::Vector{<:AbsHyperComplex}
+
+  function HCTensorProductMapFactory(
+      ::Type{MT},
+      factors::Vector{<:AbsHyperComplex}
+    ) where {MT}
+    return new{MT}(factors)
+  end
+end
+
+function zero_map(M::ModuleFP, N::ModuleFP)
+  base_ring(M) === base_ring(N) || error("base rings must coincide")
+  return hom(M, N, elem_type(N)[zero(N) for i in 1:ngens(M)], check=false)
+end
+
+function (fac::HCTensorProductMapFactory{MorphismType})(c::AbsHyperComplex, p::Int, I::Tuple) where {MorphismType}
+  next = collect(I)
+  next[p] = next[p] + (direction(c, p) == :chain ? -1 : 1)
+  J = Tuple(next)
+
+  if iszero(c[I]) || iszero(c[J])
+    return zero_map(c[I], c[J])
+  end
+  
+  i = collect(I)
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  maps = MorphismType[]
+  d = length.(j)
+  k = findfirst(k->sum(d[1:k]; init=0) >= p, 1:length(j))
+  k === nothing && error("direction out of bounds")
+  q = p - sum(d[1:k-1]; init = 0)
+  maps = MorphismType[identity_map(fac.factors[l][Tuple(j[l])]) for l in 1:k-1]
+  push!(maps, map(fac.factors[k], q, Tuple(j[k]))) 
+  maps = vcat(maps, MorphismType[identity_map(fac.factors[l][Tuple(j[l])]) for l in k+1:length(j)])
+  return tensor_product(c[I], c[J], maps)
+end
+
+function can_compute(fac::HCTensorProductMapFactory, c::AbsHyperComplex, p::Int, I::Tuple)
+  i = collect(I)
+  # decompose I into its individual indices
+  j = Vector{Vector{Int}}()
+  k = 0
+  for f in fac.factors
+    push!(j, i[k+1:k+dim(f)])
+    k = k + dim(f)
+  end
+  d = length.(j)
+  l = findfirst(l->sum(d[1:l]; init=0)>=p, 1:length(j))
+  l === nothing && error("mapping direction out of bounds")
+  return can_compute_map(fac.factors[l], p - sum(d[1:l-1]; init=0), Tuple(j[l]))
```

Comment:
It may not matter (because `length(j)` is "always" small, perhaps???) but the `findfirst` line is needlessly using a quadratic algorithm to solve a problem that's easily solvable in linear time. E.g. inside the `for` loop above. In fact, isn't `k` equal to the partial sums you are looking at here? And you are only using `j[l]` later ... So can't you do something like this:
```suggestion
  i = collect(I)
  # decompose I into its individual indices
  k = 0
  for f in fac.factors
    if k + dim(f) >= p
      r = i[k+1:k+dim(f)]
      return can_compute_map(f, p - k, Tuple(r))
    end
    k = k + dim(f)
  end
  error("mapping direction out of bounds")
```

Reply:
That seems neat. Thanks.